### ISSUE TITLE
Loop Redesign

### ISSUE THREAD

**Issue by JacquesLucke** [JacquesLucke/animation_nodes#173]
In general loops work very well if you know how they work... Unfortunally only a rare number of users understand that in detail and are often very confused by the results.

To make loops more usable I thought of different loop types. Most of the time the user either wants to create a list of something or use an existing list (of eg objects) and set parameters on this.

Here is a small idea how a loop that creates lists can look like (I used some script nodes to make real looking nodes)
![list creation prototype](https://cloud.githubusercontent.com/assets/5881326/8968898/77a6dc52-3640-11e5-994f-ca4a863cc32d.PNG)

**Comment by og76** [comment]
I also see that the current separation in lists / option is confusing,
tough I eventually found it **very convenient**
![new](https://cloud.githubusercontent.com/assets/11709974/9254255/3100f52e-41eb-11e5-8d4f-5670c65e5868.PNG)

in this regard, maybe it's enough to have a more visual separation, but separators or so between sockets, it's not possible or I don't know how to do it ... 
(unless we have a fake socket type, with 0 alpha and links limit 0, as a fake label..)

**Comment by JacquesLucke** [comment]
Well yes from the basic idea this will stay the same although I want to extend it a bit and maybe rename some stuff. Also it doesn't look great to have all the inputs as outputs atm. I want to separate them more (but the current way will work as well)

Basicly a loop can have two types as input (originally called "List" and "Option"). The first will (propably) be called "Iterator" and the other "Parameter". They serve the same functionality like list and option nowadays but won't be loop outputs by default.

Beside that a loop can have two output types (which will have output sockets only (most likely)):
1. the "Generator" output makes it easy to create a list of whatever you want. Basicly you append a custom number of elements to that list in each loop iteration. The final list will be the output in the caller node.this is what I wanted to showcase in the Screenshot above.
2. the "accumulated" output will be a value that can be changed in every loop iteration. Eg: you have a list of objects as iterator input and a accumulated integer as output. The integer starts at zero. In each loop Iteration you can add the z location of the object to that integer. In the end you will have the calculted value as loop output. 

Beside that I want to let the user choose what the iteration count defines. Currently when there is no list input there will the an Amount input. If there is one list it will take its length and if there are more than one it will take the smallest length. Now I want to take these 3 options (and also "Automatic" which will work like the behavior nowadays) and put them into a enumproperty to let the user choose.

**Comment by og76** [comment]
- "Iterator" and "Parameter" are more suggestive.
- the output nodes will make a lot of difference, as visual nodal approach (guess that is the big idea)
- I'm not sure I follow the accumulated variant...

- chosing the count (amount ) covers another issue, tied to iterator/parameter, as I think it's not clear today that the obj/poly etc options are just presets. 

Here's an idea. Maybe it would be more obvious like this:
enum to choose from (Arbitrary amount, List length, Auto)
 - for arbitrary (custom) amount, like today, but allowing lists plugged (no matter their len).
[--more like for i in range ...kind]
 - **for List length (iterator list len?) it should show extra button (check) aside respective lists (iterator) to mark which one I wanna count.**
[--like today's manage socket does with respective del/move buttons]
but what if I have no list there? a warning or so? hm
 - for Auto like today, but somehow it should show (an icon aside the list..?) 
I'll take this as a legacy /compatibility variant

![loops](https://cloud.githubusercontent.com/assets/11709974/9264910/93a50de8-4234-11e5-95f4-4edea1d8bb9b.jpg)

**Comment by JoseConseco** [comment]
About generators and accumulated outputs.  I wish they worked not as outputs, but as a special type of loop. Let me describe why.
Often when I work with AnimNodes I want to **initialize** my scene setup (**once**) and then do some animation or play with some parameters to see how they affect objects in interactive way  - it is possible when 'scene changed' or 'property changed' is enabled.
![image](https://cloud.githubusercontent.com/assets/13521338/9275157/6a98deae-429b-11e5-90cb-3e4a2e56331d.png)
With ability to initialize some parameters/ objects in "Generator" loop we could do some heavy stuff there ( create, calculate object data) once, and then all interactive work/animation could be served in "accumulated" loop (preferably simple calculations). The benefit would be - user could choose what he wants to calculate (preferably complex stuff) only once in 'generate loop', and then easy stuff could be executed the old way in 'accumulated loops'. 
This would require something like 'Play/Run' button, after pressing it all 'generator loops' would be executed, preparing data for further processing. After that ''accumulated' loop would work same as normal loop now.
Maybe I should post this in new thread, but when I saw Jacques writing about  'generators' and 'accumulated' outputs it reminded me the problem I had with calculation of heavy data over and over in each loop iteration, without way to initialize it only once. Cache partially solved this, but it was not flexible enough, and it was disabled by default. 
 
About "amount" for loop iteration count, I think the easiest way would be to leave it always visible, and require user to manually plug list length to this input. It may require 2 steps more to setup, but imo it would be easiest to understand. I had  problem with this at beginning too.

**Comment by JacquesLucke** [comment]
Good ideas o.g. I will think choose a way when I'm at this point :)
An important thing to consider is performance when we talk about loops.
Having code like: 
for i in range(...): o = objects[i]
Is much!! slower than:
for i in objects: ...
So always using the amount input would not be a good idea. The good thing about using the shortest iterator input length is that I can use the zip function in python which offers very good performance. Unfortunally it is not the easiest thing to understand for normal users.

So if I understand it right, JoseConseco, you are mainly talking about caching. I mean to calculate a list or whatever once and reuse it in the following executions.
"Play/Run" buttons won't be the way to go I think.
Instead I just got another idea for a more generic caching solution. But here are first some problems:
- not all socket types can be cached. Eg an object which can cause blender to crash when a script tries that (other types like integer are easy to cache)
- depending on the data type and for what you use it the cached data has to be copied for each use so that you don't accidently change the cache. The time to copy that can sometimes take as long as creating the cache itself.
To understand my newest caching idea you have to know that I want to unify Groups, Loops and Scripts to "Subprograms" or "Procedures" because this is what they essentially all are. Each subprogram has a unique identifier and a name which can appear multiple times (but shouldn't for usability reasons). This is important to avoid conflicts when you append a nodetree from another file which has a subprogram with a name that already exists.
The benefit is that I need only one Caller node. Maybe I can add a checkbox to that caller node that says "Cache Result". When true the subprogram will only be calculated once. After that in each execution the subprogramm will be skipped and the cached data will be used as output instead. (This only works as long as all outputs have cacheable types)
That should solve all issues with caching.. I know it is a bit hard to understand for you now but you will understand it once I have a working implementation. Please continue sharing your ideas here. This really helps to think into different directions and even if I don't use your idea in the and it still helps a lot.

Btw: when you have to calculate the same heavy data in each loop iteration you are most propably doing something wrong. Maybe you can speed it up by taking the calculation out of the loop and use the result as loop input parameter (option)

**Comment by og76** [comment]
1.**I think that normally you would use Lists (iterator) for lists of same len, for convenience**
 I don't immediately see the case where I would use different len in Lists (iterator).
Any smaller or longer lists would go into Option (parameter) and use a get element, for an operation that does not involve the normal index.

ex: 3 objects to random place n times, the yellow nodes take care of the different list.
![distribute random](https://cloud.githubusercontent.com/assets/11709974/9279236/cebaa12e-42be-11e5-8d85-e56dfd298488.PNG)

however, if I can check the list that I wanna get the index from, it's all clear

2.There must be a way instead of *i in range*
-have an integer list [0-amount] and perform same kind of zip (map ?) operation
-arrays?

and so on


(I talk out of shallow reading ... )

**Comment by JacquesLucke** [comment]
Youre right. Normally multiple iterator inputs should have the same length but sometimes (eg due to user failure) they don't have the same. However I can think of examples where this actually would be useful but only in rare cases. 

2. hmm I don't know what you mean...

**Comment by JoseConseco** [comment]
Jacques, about "not all socket types can be cached. Eg an object which can cause blender to crash when a script tries that", maybe I can report bug to blender bugtracker. But I would need better description. Or you could fill in bug in here yourself, when you are back:
https://developer.blender.org/maniphest/

**Comment by JacquesLucke** [comment]
This is not considered as a real bug by the institute. They even explain why you shouldn't store a blender object over a longer period of time in the api reference. I don't have a link atm. Basicly it is all about better performance for script execution
